gcc/cp/contracts.cc

```diff
@@ -3464,150 +3394,246 @@ p2900_duplicate_contracts (tree newdecl, tree olddecl)
-  tree old_contracts = DECL_CONTRACTS (olddecl);
-  tree new_contracts = DECL_CONTRACTS (newdecl);
+  tree old_contracts = DECL_CONTRACTS(olddecl);
+  tree new_contracts = DECL_CONTRACTS(newdecl);
 
   if (!old_contracts && !new_contracts)
     return;
 
-  location_t old_loc = DECL_SOURCE_LOCATION (olddecl);
-  location_t new_loc = DECL_SOURCE_LOCATION (newdecl);
+  location_t old_loc = DECL_SOURCE_LOCATION(olddecl);
+  location_t new_loc = DECL_SOURCE_LOCATION(newdecl);
 
   /* If both declarations specify contracts, ensure they match.
 
-     TODO: This handles a potential error a little oddly. Consider:
+   TODO: This handles a potential error a little oddly. Consider:
 
-	struct B {
-	  virtual void f(int n) [[pre: n == 0]];
-	};
-	struct D : B {
-	  void f(int n) override; // inherits contracts
-	};
-	void D::f(int n) [[pre: n == 0]] // OK
-	{ }
+   struct B {
+   virtual void f(int n) [[pre: n == 0]];
+   };
+   struct D : B {
+   void f(int n) override; // inherits contracts
+   };
+   void D::f(int n) [[pre: n == 0]] // OK
+   { }
 
-    It's okay because we're explicitly restating the inherited contract.
-    Changing the precondition on the definition D::f causes match_contracts
-    to complain about the mismatch.
+   It's okay because we're explicitly restating the inherited contract.
+   Changing the precondition on the definition D::f causes match_contracts
+   to complain about the mismatch.
 
-    This would previously have been diagnosed as adding contracts to an
-    override, but this seems like it should be well-formed.  */
+   This would previously have been diagnosed as adding contracts to an
+   override, but this seems like it should be well-formed.  */
   if (old_contracts && new_contracts)
     {
-      if (!match_contract_conditions (old_loc, old_contracts,
-				      new_loc, new_contracts,
-				      cmc_declaration))
-	return;
-      if (DECL_UNIQUE_FRIEND_P (newdecl))
-	/* Newdecl's contracts are still DEFERRED_PARSE, and we're about to
-	   collapse it into olddecl, so stash away olddecl's contracts for
-	   later comparison.  */
-	defer_guarded_contract_match (olddecl, olddecl, old_contracts);
-      remove_contract_attributes (olddecl);
+      if (!contract_any_deferred_p (old_contracts)
+	  && contract_any_deferred_p (new_contracts)
+	  && DECL_UNIQUE_FRIEND_P (newdecl))
+	{
+	  /* The contracts of newdecl are still DEFERRED_PARSE, and we're about
+	     to collapse it into olddecl, so stash away olddecl's contracts for
+	     later comparison.  */
+	  defer_guarded_contract_match (olddecl, olddecl, old_contracts);
+	  /* Put the defered contracts on the olddecl so we parse it when
+	     we can.  */
+	  copy_deferred_contracts(newdecl, olddecl);
+	}
+      else if (contract_any_deferred_p (old_contracts)
+	  || contract_any_deferred_p (new_contracts))
+	  {
+	  /* TODO: ignore these and figure out how to process them later.  */
+	  /* Note that a in class friend declaration has deferred contracts, but out
```

Comment:
This needs the friend-phrasing improvement, too.

Reply:
done